src/basic/mkdir.c

```diff
@@ -248,16 +251,25 @@ int mkdir_p_root(const char *root, const char *p, uid_t uid, gid_t gid, mode_t m
                 return r;
         }
 
-        if (uid_is_valid(uid) || gid_is_valid(gid)) {
-                _cleanup_close_ int nfd = -EBADF;
+        if (ts == USEC_INFINITY && !uid_is_valid(uid) && !gid_is_valid(gid)) {
+                return 1;
+        }
 
+        if (ts != USEC_INFINITY) {
                 nfd = openat(dfd, bn, O_RDONLY|O_CLOEXEC|O_DIRECTORY|O_NOFOLLOW);
                 if (nfd < 0)
                         return -errno;
 
-                if (fchown(nfd, uid, gid) < 0)
+                timespec_store_nsec(&tspec, ts);
+                if (futimens(dfd, (const struct timespec[2]) { { .tv_nsec = UTIME_OMIT }, tspec }) < 0)
+                        return -errno;
+
+                if (futimens(nfd, (const struct timespec[2]) { tspec, tspec }) < 0)
                         return -errno;
         }
 
+        if ((uid_is_valid(uid) || gid_is_valid(gid)) && (fchown(nfd, uid, gid) < 0))
+                return -errno;
```

Comment:
this will fall apart if uid or gid are specified but no timestamp, because nfd will not be initialized then.